internal/handler/msg_aggregate.go

```diff
@@ -248,10 +248,21 @@ func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 	}
 
 	cancel := func() {}
+
+	timeout := time.NewTicker(time.Duration(maxTimeMS) * time.Millisecond)
+	findDone := make(chan struct{})
+
 	if maxTimeMS != 0 {
-		// It is not clear if maxTimeMS affects only aggregate, or both aggregate and getMore (as the current code does).
-		// TODO https://github.com/FerretDB/FerretDB/issues/2983
-		ctx, cancel = context.WithTimeout(ctx, time.Duration(maxTimeMS)*time.Millisecond)
+		ctx, cancel = context.WithCancel(ctx)
+
+		go func() {
+			select {
+			case <-timeout.C:
+				cancel()
+			case <-findDone:
+				return
```

Comment:
Maybe use a `case <-ctx.Done()`?

Reply:
In which case? Those lines are managing the context cancellation, not waiting for it

Reply:
Oh, I see. I was mostly worried about the case when the function returns due to an error and `findDone <- struct{}{}` doesn't happen.